src/libexpr/primops.cc

```diff
@@ -3712,16 +3712,27 @@ static RegisterPrimOp primop_toString({
 static void prim_substring(EvalState & state, const PosIdx pos, Value * * args, Value & v)
 {
     int start = state.forceInt(*args[0], pos, "while evaluating the first argument (the start offset) passed to builtins.substring");
-    int len = state.forceInt(*args[1], pos, "while evaluating the second argument (the substring length) passed to builtins.substring");
-    NixStringContext context;
-    auto s = state.coerceToString(pos, *args[2], context, "while evaluating the third argument (the string) passed to builtins.substring");
 
     if (start < 0)
         state.debugThrowLastTrace(EvalError({
             .msg = hintfmt("negative start position in 'substring'"),
             .errPos = state.positions[pos]
         }));
 
+
+    int len = state.forceInt(*args[1], pos, "while evaluating the second argument (the substring length) passed to builtins.substring");
+
+    if (len <= 0) {
+        state.forceValue(*args[2], pos);
```

Comment:
Not required, but I'd be ok with loosening up strictness as well, in a second commit, if you add a test and a release note.
```suggestion
```

You could test it with something like
```
builtins.substring 2 2 (throw "string must not be forced when the range is empty")
```
in the `tests/functional/lang/eval-ok-*` suite.

Reply:
The first version I pushed had an error, where I wasn't adding the string context correctly. If arg[2] is a thunk we have to force it be able to get the context out, I think?

Reply:
I don't think we need a release note if something becomes *less* strict, since that shouldn't break anything. Might be worth mentioning in the docs for `substring`, though.

Reply:
I think we do want `substring` to check if its argument is a string. Otherwise weird type errors could go uncaught.

Reply:
> we have to force it be able to get the context out, I think?

Correct. That's important behavior that I overlooked, so my suggestion here was wrong.